tequilapi/sso/mystnodes.go

```diff
@@ -0,0 +1,181 @@
+	"github.com/mysteriumnetwork/node/requests"
+	"github.com/mysteriumnetwork/node/tequilapi/contract"
+	"github.com/mysteriumnetwork/node/tequilapi/pkce"
+	"github.com/pkg/errors"
+)
+
+// ErrHostMissing host can't be blank
+var ErrHostMissing = errors.New("host must not be empty")
+
+// ErrNoUnlockedIdentity no unlocked identity
+var ErrNoUnlockedIdentity = errors.New("lastUnlockedIdentity must not be empty")
+
+// ErrCodeVerifierMissing code verifier is missing
+var ErrCodeVerifierMissing = errors.New("no code verifier generated")
+
+// ErrAuthorizationGrantTokenMissing blank authorization token
+var ErrAuthorizationGrantTokenMissing = errors.New("token must be set")
+
+// ErrMystnodesAuthorizationFail authorization failed against mystnodes
+var ErrMystnodesAuthorizationFail = errors.New("mystnodes SSO grant authorization verification failed")
+
+type httpClient interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
+// Mystnodes SSO support
+type Mystnodes struct {
+	baseUrl                   string
+	ssoPath                   string
+	signer                    identity.SignerFactory
+	lastUnlockedIdentity      identity.Identity
+	client                    httpClient
+	lastCodeVerifierBase64url string
+	lock                      sync.Mutex
+}
+
+// NewMystnodes constructor
+func NewMystnodes(signer identity.SignerFactory, client httpClient) *Mystnodes {
+	return &Mystnodes{
+		baseUrl: config.GetString(config.FlagMMNAddress),
+		ssoPath: "/login-sso",
+		signer:  signer,
+		client:  client,
+	}
+}
+
+// Subscribe unlocked identity is required in order to sign request
+func (m *Mystnodes) Subscribe(eventBus eventbus.EventBus) error {
+	if err := eventBus.SubscribeAsync(identity.AppTopicIdentityUnlock, m.onIdentityUnlocked); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (m *Mystnodes) onIdentityUnlocked(ev identity.AppEventIdentityUnlock) {
+	m.lastUnlockedIdentity = ev.ID
+}
+
+func (m *Mystnodes) message(info pkce.Info, redirectURL string) MystnodesMessage {
+	return MystnodesMessage{
+		CodeChallenge: info.CodeChallenge,
+		Identity:      m.lastUnlockedIdentity.Address,
+		RedirectURL:   redirectURL,
+	}
+}
+
+func (m *Mystnodes) sign(msg []byte) (identity.Signature, error) {
+	return m.signer(m.lastUnlockedIdentity).Sign(msg)
+}
+
+// SSOLink build SSO link to begin authentication via mystnodes.com
+func (m *Mystnodes) SSOLink(host string) (*url.URL, error) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	if len(host) == 0 {
+		return nil, ErrHostMissing
+	}
+
+	if len(m.lastUnlockedIdentity.Address) == 0 {
+		return nil, ErrNoUnlockedIdentity
+	}
+
+	u, err := url.Parse(m.baseUrl)
+	if err != nil {
+		return &url.URL{}, err
+	}
+
+	u = u.JoinPath(m.ssoPath)
+
+	info, err := pkce.New(128)
+	if err != nil {
+		return nil, err
+	}
+
+	m.lastCodeVerifierBase64url = info.Base64URLCodeVerifier()
+	messageJson, err := m.message(info, strings.Join([]string{"http://", host, "/#/auth-sso"}, "")).json()
```

Comment:
this would also be avoided by getting redirect url from the parameter as it would already contain the http or https prefix